README.md

```diff
@@ -134,30 +151,167 @@ or in the case of an error:
+provided. If provided, the public key will be registered for all of these
+chains. If no chain IDs are provided, the public key will only be registered for
+the chain used to sign the request.
 
 The nonce from the latest GET request must be provided to prevent replay
 attacks. It starts at 0, and the GET request will return an empty profile with a
-nonce of 0 if nothing has been set. The name must be unique, at least 1
-character long, and at most 32 characters long. The NFT must be owned by the
-public key, and the signature must be made by the same public key. If the NFT
-has no image, it will fail.
+nonce of 0 if nothing has been set.
+
+The signature is derived by calling `OfflineAminoSigner`'s `signAmino` function
+with the `signDoc` argument generated using `makeSignDoc` from the
+`@cosmjs/amino` package. This can be seen in the signature verification code
+located in [src/index.ts](./src/utils/auth.ts#L20) around line 20.
+
+### `POST /register`
+
+The expected request body type is:
+
+```ts
+type RegisterPublicKeyRequest = {
+  data: {
+    publicKeys: {
+      data: {
+        allow: string;
+        chainIds?: string[];
+        auth: {
+          type: string;
+          nonce: number;
+          chainId: string;
+          chainFeeDenom: string;
+          chainBech32Prefix: string;
+          publicKey: string;
+        };
+      };
+      signature: string;
+    }[];
+    auth: {
+      type: string;
+      nonce: number;
+      chainId: string;
+      chainFeeDenom: string;
+      chainBech32Prefix: string;
+      publicKey: string;
+    };
+  };
+  signature: string;
+};
+```
+
+The returned type is:
+
+```ts
+type RegisterPublicKeyResponse = {
+  success: true;
+};
+```
+
+or in the case of an error:
+
+```ts
+type RegisterPublicKeyResponse = {
+  error: string;
+};
+```
+
+This route lets the user register public keys for their profile and/or set chain
+preferences for their registered public keys. The keys being registered must
+sign allowances (using the same nonce) to allow the profile to register them (or
+else anyone could claim any public key as their own). Thus, the keys being
+registered must set `allow` to the profile's public key that is registering
+them. `chainIds` is an optional array of chain IDs to set as preferences for
+this public key. If `chainIds` is omitted for a key being registered, it will
+just set the chain preference for the chain used to sign the nested request. The
+authentication structure is nested, so first the keys being registered need to
+sign the data objects, and then the profile public key performing the
+registration needs to sign all of them.
+
+All public keys attached to a profile will resolve to the profile when looked up
+using the GET request at the top (by address, public key, or bech32 hash).
+However, **a profile must explicitly choose a public key to use for name
+resolution/searching on each chain**. This is because chains sometimes use
```

Comment:
Interesting. Is this why we have to set a chain preference? Or is there a different reason for that?

Reply:
yep this is what a chain preference is